cli/src/composer.rs

```diff
@@ -0,0 +1,211 @@
+                params.light_props = Some(res);
+                return compose_joint(params);
+            }
+            Err(e) => bail!("request from light vendor, err = {}", e),
+        }
+    }
+    //TODO: trySubset()
+
+    let mut signing_addresses = params.signing_addresses;
+    let mut paying_addresses = params.paying_addresses;
+    let outputs = params.outputs;
+    let mut messages = params.messages.clone();
+
+    let light_props = params.light_props;
+    let signer = params.signer;
+
+    if light_props.is_none() {
+        bail!("no parent props for light");
+    }
+
+    let change_outputs = outputs
+        .iter()
+        .filter(|output| output.amount == Some(0))
+        .cloned()
+        .collect::<Vec<_>>();
+    let external_outputs = outputs
+        .iter()
+        .filter(|output| output.amount > Some(0))
+        .collect::<Vec<_>>();
+    if change_outputs.len() > 1 {
+        bail!("more than one change output");
+    }
+    if change_outputs.is_empty() {
+        bail!("no change outputs");
+    }
+    if paying_addresses.is_empty() {
+        bail!("no payers?");
+    }
+
+    let from_addresses = {
+        signing_addresses.append(&mut paying_addresses);
+        signing_addresses.sort();
+        signing_addresses
+    };
+    let hash_placeholder = "--------------------------------------------".to_string();
+    let payment_message = Message {
+        app: "payment".to_string(),
+        payload_location: "inline".to_string(),
+        payload_hash: hash_placeholder,
+        payload: Some(Payload::Payment(Payment {
+            address: None,
+            asset: None,
+            definition_chash: None,
+            denomination: None,
+            inputs: Vec::new(),
+            outputs: change_outputs.clone(),
+        })),
+        payload_uri: None,
+        payload_uri_hash: None,
+        spend_proofs: Vec::new(),
+    };
+
+    let mut total_amount = 0;
+
+    for output in external_outputs.iter() {
+        match payment_message.clone().payload.unwrap() {
+            Payload::Payment(mut x) => x.outputs.push(output.clone().clone()),
+            _ => {}
+        }
+        total_amount += output.amount.unwrap();
+    }
+
+    messages.push(payment_message);
+
+    let is_multi_authored = from_addresses.len() > 1;
+    let mut unit = spec::Unit::default();
+    unit.messages = messages.clone();
+    if is_multi_authored {
+        unit.earned_headers_commission_recipients
+            .push(spec::HeaderCommissionShare {
+                address: change_outputs.into_iter().nth(0).unwrap().address.unwrap(),
+                earned_headers_commission_share: 100,
+            });
+    }
+
+    let db = db::DB_POOL.get_connection();
+
+    // TODO: lock
+
+    // parent units
+    unit.parent_units = light_props.clone().unwrap().parent_units;
+    unit.last_ball = light_props.clone().unwrap().last_stable_mc_ball;
+    unit.last_ball_unit = light_props.clone().unwrap().last_stable_mc_ball_unit;
```

Comment:
better to have a `let props = light_props.clone().unwrap()` first and remove unnecessary clones